File: sdk/go/provider/xkx_me.go
package provider

import (
	"encoding/json"
	"fmt"
	"io"
	"net/url"
	"regexp"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

const xkxMeBase = "https://xkx.me"

var xkxMeCsrfRe = regexp.MustCompile(`csrf-token" content="([^"]+)"`)

type xkxMeSession struct {
	csrf    string
	cookies string
}

func xkxMeGetSession() (*xkxMeSession, error) {
	client := HTTPClient()
	req, err := http.NewRequest(http.MethodGet, xkxMeBase, nil)
	if err != nil {
		return nil, err
	}
	ua := getCurrentUA()
	if ua != "" {
		req.Header.Set("User-Agent", ua)
	}
	req.Header.Set("Accept", "text/html")

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	m := xkxMeCsrfRe.FindStringSubmatch(string(body))
	if len(m) < 2 {
		return nil, fmt.Errorf("xkx-me: 无法获取 CSRF token")
	}

	var cookieParts []string
	for _, c := range resp.Cookies() {
		cookieParts = append(cookieParts, c.Name+"="+c.Value)
	}

	return &xkxMeSession{
		csrf:    m[1],
		cookies: strings.Join(cookieParts, "; "),
	}, nil
}

var xkxMeEmailRe = regexp.MustCompile(`mailbox/([^\s/"'<>]+@xkx\.me)`)

func XkxMeGenerate() (*CreatedMailbox, error) {
	session, err := xkxMeGetSession()
	if err != nil {
		return nil, err
	}

	client := HTTPClient()
	formData := "_token=" + url.QueryEscape(session.csrf)
	req, err := http.NewRequest(http.MethodPost, xkxMeBase+"/mailbox/create/random", strings.NewReader(formData))
	if err != nil {
		return nil, err
	}

	ua := getCurrentUA()
	if ua != "" {
		req.Header.Set("User-Agent", ua)
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Cookie", session.cookies)

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	location := resp.Header.Get("Location")
	if location == "" {
		b, _ := io.ReadAll(resp.Body)
		location = string(b)
	}

	em := xkxMeEmailRe.FindStringSubmatch(location)
	if len(em) < 2 {
		return nil, fmt.Errorf("xkx-me: 无法从响应中提取邮箱地址")
	}

	return &CreatedMailbox{
		Channel: "xkx-me",
		Email:   em[1],
		Token:   session.cookies,
	}, nil
}

func XkxMeGetEmails(token, email string) ([]NormEmail, error) {
	address := strings.TrimSpace(email)
	if address == "" {
		return nil, fmt.Errorf("xkx-me: 缺少邮箱地址")
	}

	client := HTTPClient()
	u := fmt.Sprintf("%s/mailbox/%s/messages", xkxMeBase, url.PathEscape(address))
	req, err := http.NewRequest(http.MethodGet, u, nil)
	if err != nil {
		return nil, err
	}

	ua := getCurrentUA()
	if ua != "" {
		req.Header.Set("User-Agent", ua)
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("X-Requested-With", "XMLHttpRequest")
	if token != "" {
		req.Header.Set("Cookie", token)
	}

	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode == 404 {
		return []NormEmail{}, nil
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("xkx-me: 获取邮件失败 HTTP %d", resp.StatusCode)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	var messages []map[string]any
	if err := json.Unmarshal(body, &messages); err != nil {
		var wrapper struct {
			Messages []map[string]any `json:"messages"`
		}
		if err2 := json.Unmarshal(body, &wrapper); err2 != nil {
			return []NormEmail{}, nil
		}
		messages = wrapper.Messages
	}

	if len(messages) == 0 {
		return []NormEmail{}, nil
	}

	out := make([]NormEmail, 0, len(messages))
	for _, msg := range messages {
		htmlContent := xkxMeToString(msg["html"])
		if htmlContent == "" {
			htmlContent = xkxMeToString(msg["body"])
		}
		raw := map[string]any{
			"id":      msg["id"],
			"from":    msg["from"],
			"to":      address,
			"subject": msg["subject"],
			"date":    msg["date"],
			"html":    htmlContent,
			"text":    msg["text"],
		}
		out = append(out, NormalizeMap(raw, address))
	}
	return out, nil
}

func xkxMeToString(v any) string {
	if v == nil {
		return ""
	}
	if s, ok := v.(string); ok {
		return s
	}
	return fmt.Sprintf("%v", v)
}

File: sdk/npm/src/providers/xkx-me.ts
import { InternalEmailInfo, Email, Channel } from "../types";
import { normalizeEmail } from "../normalize";
import { fetchWithTimeout } from "../retry";

const CHANNEL: Channel = "xkx-me";
const BASE_URL = "https://xkx.me";

/**
 * 从首页 HTML 提取 CSRF token
 */
function extractCsrf(html: string): string {
  const m = html.match(/csrf-token" content="([^"]+)"/);
  if (!m) throw new Error("xkx-me: 无法获取 CSRF token");
  return m[1];
}

/**
 * 从 Set-Cookie 头中提取 cookies
 */
function extractCookies(resp: Response): string {
  const raw = resp.headers.get("set-cookie") || "";
  if (!raw) return "";
  return raw
    .split(",")
    .map((c) => c.split(";")[0].trim())
    .filter(Boolean)
    .join("; ");
}

/**
 * 创建 xkx.me 临时邮箱
 * 1. GET / 获取 CSRF token 和 session cookie
 * 2. POST /mailbox/create/random 创建随机邮箱
 */
export async function generateEmail(): Promise<InternalEmailInfo> {
  const pageResp = await fetchWithTimeout(BASE_URL, {
    method: "GET",
    headers: {
      Accept: "text/html",
      "User-Agent":
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
    },
  });
  if (!pageResp.ok) {
    throw new Error(`xkx-me: 获取首页失败 HTTP ${pageResp.status}`);
  }

  const pageHtml = await pageResp.text();
  const csrf = extractCsrf(pageHtml);
  const cookies = extractCookies(pageResp);

  const createResp = await fetchWithTimeout(`${BASE_URL}/mailbox/create/random`, {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
      Cookie: cookies,
      "User-Agent":
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
    },
    body: `_token=${encodeURIComponent(csrf)}`,
    redirect: "manual",
  });

  const location = createResp.headers.get("location") || "";
  let body = "";
  if (!location) {
    body = await createResp.text();
  }

  const source = location || body;
  const emailMatch = source.match(/mailbox\/([^\s/"'<>]+@xkx\.me)/);
  if (!emailMatch) {
    throw new Error("xkx-me: 无法从响应中提取邮箱地址");
  }

  return {
    channel: CHANNEL,
    email: emailMatch[1],
    token: cookies,
  };
}

/**
 * 获取 xkx.me 邮件列表
 * GET /mailbox/{email}/messages (JSON)
 */
export async function getEmails(
  token: string,
  email: string,
): Promise<Email[]> {
  const address = String(email || "").trim();
  if (!address) {
    throw new Error("xkx-me: 邮箱地址为空");
  }

  const resp = await fetchWithTimeout(
    `${BASE_URL}/mailbox/${encodeURIComponent(address)}/messages`,
    {
      method: "GET",
      headers: {
        Accept: "application/json",
        "X-Requested-With": "XMLHttpRequest",
        Cookie: token || "",
        "User-Agent":
          "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
      },
    },
  );

  if (resp.status === 404) return [];
  if (!resp.ok) {
    throw new Error(`xkx-me: 获取邮件失败 HTTP ${resp.status}`);
  }

  const data = await resp.json();
  const messages: Array<Record<string, unknown>> = Array.isArray(data)
    ? data
    : data?.messages || [];

  if (messages.length === 0) return [];

  return messages.map((msg) => {
    const raw: Record<string, unknown> = {
      id: msg.id,
      from: msg.from,
      to: address,
      subject: msg.subject,
      date: msg.date,
      html: msg.html || msg.body || "",
      text: msg.text || "",
    };
    return normalizeEmail(raw, address);
  });
}
